### PR TITLE
Added Session 10 Assignment

### PR DESCRIPTION
**Pull request by jagadish-varanasi**

### CODE REVIEW

File: session-10/index.html [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Document</title>
+  </head>
+  <body>
+    <div id="root"></div>
+    <script type="module" src="./src/App.js"></script>
```
Comment (ap221882) [inline session-10/index.html:11]: modules are deferred by default, can put it in head tag

File: session-10/src/App.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import "./App.css";
+import { lazy, Suspense } from "react";
+import ReactDOM from "react-dom/client";
+import { createBrowserRouter, RouterProvider } from "react-router-dom";
+import { UserContextProvider } from "./context/UserContext/UserContext";
+import { ThemeContextProvider } from "./context/ThemeContext/ThemeContext";
```
Comment (ap221882) [inline session-10/src/App.js:6]: Keep global imports at the top, then keep local imports below it. Lazy imports can be at the bottom. This means keeping your imports also clean and understandable.

File: session-10/src/App.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+const Home = lazy(() => import("./pages/Home/Home"));
```
Comment (ap221882) [inline session-10/src/App.js:7]: This(Home) is your default route. No need to import it lazily, or give me the reason if there is any.

File: session-10/src/components/Header/Header.css [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+.header-dark {
+  background-color: #ffff;
+  padding: 1rem;
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+}
+
+input,
```
Comment (ap221882) [inline session-10/src/components/Header/Header.css:22]: Keep your body, *, and html-tag CSS selectors on top, keeping in mind CSS specificity.

File: session-10/src/components/Header/Header.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import { useThemeContext } from "../../context/ThemeContext/ThemeContext";
+
+function Header() {
+  const { theme, setTheme } = useThemeContext();
+  const { reference, handleSearchText } = useUserContext();
+
+  const themeButton = theme == "light" ? "Dark" : "Light";
+
+  const toggleTheme = () => {
+    setTheme(theme == "light" ? "dark" : "light");
```
Comment (ap221882) [inline session-10/src/components/Header/Header.js:15]: make variables of `theme === "light"` and `theme === "dark"` on top. Use them throughout your code.

File: session-10/src/components/Header/Header.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  };
+
+  return (
+    <div className={theme == "dark" ? "header-dark" : "header-light"}>
+      <img src={logo} alt="logo" width="40px" height="40px" />
```
Comment (ap221882) [inline session-10/src/components/Header/Header.js:20]: avoid inline styling, is less readable

File: session-10/src/components/SelectBox/SelectBox.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+  return (
+    <div className="select">
+      <select
+        name="name"
+        value={selectedState}
+        onChange={(e) => setSelectedState(e.target.value)}
+      >
+        {states.map((state) => (
+          <option value={state}>{state}</option>
```
Comment (ap221882) [inline session-10/src/components/SelectBox/SelectBox.js:25]: Where is the key prop when you're using map?

File: session-10/src/components/card/Card.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import { useThemeContext } from "../../context/ThemeContext/ThemeContext";
+import "./Card.css";
+
+const Card = ({ member: { login, avatar_url, html_url } }) => {
+  const { theme } = useThemeContext();
```
Comment (ap221882) [inline session-10/src/components/card/Card.js:6]: Keep a blank line before return for readability.

File: session-10/src/context/ThemeContext/ThemeContext.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import { createContext, useContext, useState } from "react";
+
+const ThemeContext = createContext();
+
+export const ThemeContextProvider = ({ children }) => {
+  const [theme, setTheme] = useState("light");
+  return (
+    <ThemeContext.Provider value={{ theme, setTheme }}>
```
Comment (ap221882) [inline session-10/src/context/ThemeContext/ThemeContext.js:8]: Kudos!

File: session-10/src/components/Header/Header.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import React, { useRef, useState } from "react";
```
Comment (ap221882) [inline session-10/src/components/Header/Header.js:1]: remove useRef as you've moved it to the hook

File: session-10/src/context/UserContext/UserContext.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import { createContext, useContext, useState, useRef } from "react";
+import useGitHubMembersByCity from "../../hooks/useGithubMembersByCity";
+
+const UserContext = createContext();
+
+export const UserContextProvider = ({ children }) => {
+  const ref = useRef();
```
Comment (ap221882) [inline session-10/src/context/UserContext/UserContext.js:7]: Kudos!

File: session-10/src/pages/Home/Home.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+              if (searchText === "") return member;
+              else if (
+                (member.name ?? member.login)
+                  .toLocaleLowerCase()
+                  .includes(searchText.toLocaleLowerCase())
+              ) {
+                return member;
+              }
+            })
+            .map((member) => <Card key={member.login} member={member} />)}
```
Comment (ap221882) [inline session-10/src/pages/Home/Home.js:33]: If no user is found, show the user that no user is found with that search parameter

File: session-9/src/pages/Home/Home.js [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  const { data, loading } = useGitHubMembersByCity(selectedCity);
+
+  console.log(data);
+
+  const handleSearchText = () => {
+    setSearchText(ref.current.value);
+  };
+
+  return (
+    <Fragment>
```
Comment (ap221882) [inline session-9/src/pages/Home/Home.js:24]: I like `<></>` more than Fragment. Reduces one import and less code writing.

File: session-9/src/pages/Home/Home.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+      <Header reference={ref} handleSearchText={handleSearchText} />
+      <SelectBox
+        selectedState={selectedState}
+        setSelectedState={setSelectedState}
+        selectedCity={selectedCity}
+        setSelectedCity={setSelectedCity}
+      />
+      <div className="card-list">
+        {loading && <div>Loading...</div>}
```
Comment (ap221882) [inline session-9/src/pages/Home/Home.js:33]: set the loading state to false where it should be and remove it from Line 34. Can also use the ternary operator.

File: session-9/src/hooks/useGithubMembersByCity.js [hunk 15]
```diff
@@ -0,0 +1,22 @@
+import { useEffect, useState } from "react";
+
+const useGitHubMembersByCity = (selectedCity) => {
+  const [loading, setLoading] = useState(false);
+  const [data, setData] = useState([]);
+  useEffect(() => {
+    console.log("Hello");
+    (async () => {
+      setLoading(true);
+      const response = await fetch(
```
Comment (ap221882) [inline session-9/src/hooks/useGithubMembersByCity.js:10]: can move this fetching logic into different function or module.

File: session-9/src/hooks/useGithubMembersByCity.js [hunk 16]
```diff
@@ -0,0 +1,22 @@
+import { useEffect, useState } from "react";
+
+const useGitHubMembersByCity = (selectedCity) => {
+  const [loading, setLoading] = useState(false);
+  const [data, setData] = useState([]);
+  useEffect(() => {
+    console.log("Hello");
+    (async () => {
+      setLoading(true);
```
Comment (ap221882) [inline session-9/src/hooks/useGithubMembersByCity.js:9]: use always try-catch inside the async function, or what is the use of the async function then?

**Review by ap221882 — COMMENTED** [review]
Very neatly written and well-composed. Added comments. Kudos!

File: session-10/src/App.js (hunk 3 above)
Comment (jagadish-varanasi) [reply session-10/src/App.js:7]: Yes right, here Home need not to be lazily loaded as it's initial component. Changing this.

File: session-9/src/hooks/useGithubMembersByCity.js (hunk 16 above)
Comment (jagadish-varanasi) [reply session-9/src/hooks/useGithubMembersByCity.js:9]: Yes it's right always good to use try/catch for handling server issues. But I guess still this function need to async as it need to wait for promise to resolve. wrapping this function with try/catch.

**Comment by jagadish-varanasi** [comment]
> Very neatly written and well-composed. Added comments. Kudos!

Thank you for reviewing and providing feedback.

File: session-9/src/hooks/useGithubMembersByCity.js (hunk 16 above)
Comment (ap221882) [reply session-9/src/hooks/useGithubMembersByCity.js:9]: Don't wrap it inside try-catch, use try-catch inside it.

```
(async () => {
      try{
      setLoading(true);
      const response = await fetch(
        `https://api.github.com/search/users?q=location:${selectedCity}`
      );
      const data = await response.json();
      setData(data);
      } catch (error){
      console.log(error)
      }finally{
      setLoading(false);
      })()
```

Comment (jagadish-varanasi) [reply session-9/src/hooks/useGithubMembersByCity.js:9]: Yes right, But still it does bubble out right till it gets caught.😀

Comment (ap221882) [reply session-9/src/hooks/useGithubMembersByCity.js:9]: You didn't get it.

**Pull request merged**